x-pack/apm-server/sampling/pubsub/pubsub.go

```diff
@@ -87,23 +88,43 @@ func (p *Pubsub) onBulkIndexerItemFailure(ctx context.Context, item elasticsearc
 	p.config.Logger.With(logp.Error(err)).Debug("publishing sampled trace ID failed", resp.Error)
 }
 
-// SubscribeSampledTraceIDs subscribes to new sampled trace IDs, sending them to the
-// traceIDs channel.
-func (p *Pubsub) SubscribeSampledTraceIDs(ctx context.Context, traceIDs chan<- string) error {
+// SubscribeSampledTraceIDs subscribes to sampled trace IDs after the given position,
+// sending them to the traceIDs channel, and sending the most recently observed position
+// (on change) to the positions channel.
+func (p *Pubsub) SubscribeSampledTraceIDs(
+	ctx context.Context,
+	pos SubscriberPosition,
+	traceIDs chan<- string,
+	positions chan<- SubscriberPosition,
+) error {
 	ticker := time.NewTicker(p.config.SearchInterval)
 	defer ticker.Stop()
 
-	observedSeqnos := make(map[string]int64)
+	// Only send positions on change.
+	var positionsOut chan<- SubscriberPosition
+	positionsOut = positions
+
+	// Copy pos because it may be mutated by p.searchTraceIDs.
+	pos = copyPosition(pos)
 	for {
 		select {
 		case <-ctx.Done():
 			return ctx.Err()
+		case positionsOut <- pos:
+			// Copy pos because it may be mutated by p.searchTraceIDs.
+			pos = copyPosition(pos)
```

Comment:
> we mutate pos in p.searchTraceIDs, and then update the channel listener by sending a copy of the mutated pos

We mutate pos in p.searchTraceIDs, send pos to the channel, and _then_ copy it. This ensures the position sent over the channel won't be mutated.

> Is it avoiding a potential race between sending the position and having it marshaled/written to disk, and updating the value at the next tick?

Yes, the reason is to avoid a data race between the receiver of the `positions` channel (writing to disk) and `p.searchTraceIDs`. We need to make sure the map inside the position isn't modified after it has been sent over the channel.

Reply:
Gonna merge, let me know if you have any concerns later.